Add a modified osqp_solver that saves the osqp object

yangwill opened the pull request:
This solver uses osqp but saves the osqp object and corresponding data objects to prevent unnecessary work constructing and destructing.

It does this by adding a function `InitializeSolver` which creates the `OSQPWorkspace` object. Provided the structure of the mathematical program does not change, calling `Solve` will now only update the costs and constraints. This along with the now-enabled native warm starting of `osqp`, speeds up the solve time significantly.

Testing on my lab desktop with the simulator running at 2000Hz, (so max controller rate is 2000Hz):
Controller frequency of the walking controller: ~800Hz (drake osqp_solver), ~1200Hz (this PR)
Controller frequency of the standing controller: ~1450Hz (drake osqp_solver), ~1700Hz (this PR)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dairlab/dairlib/244)
<!-- Reviewable:end -->

yminchen reviewed (commented):
Code looks good. Just one question -- Do you think we can make this change in Drake? I just diffed `OsqpSolver` and `FastOsqpSolver` and it seems that the difference is not big. 
We should probably talk to Hongkai, but I think we can extend Drake's OsqpSolver by giving the constructor a flag (with a default value) that determines whether or not we intend to initialize `OSQPWorkspace`. Then inside `DoSolve`, if the flag is true, we update the values of `OSQPWorkspace`. Otherwise, we construct a new one. 
The advantage of making a PR in Drake will be that the developers will maintain the code for us when they make changes to Drake. 

> Reviewed 29 of 37 files at r3, 8 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/244#-:-MZovwsk5oY-hCMPA8aq:buowins)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @yangwill)



<!-- Sent from Reviewable.io -->

yminchen reviewed (commented):
Also 50% speed-up is significant, so I think both Drake developers and users will be happy to have this feature 

> *[Reviewable](https://reviewable.io/reviews/dairlab/dairlib/244#-:-MZp0Xe49sNhtWAO1nJw:b-krd8wv)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @yangwill)



<!-- Sent from Reviewable.io -->

yangwill commented:
Added joint limits in the OSC, ready for final review.

yangwill commented:
Summary of the changes in this branch
- Modified OSQPSolver that stores the osqp workspace and warmstarts the solve
- Added a gflag to the state estimator for the ontact force threshold
- Added joint limits as a cost in the OSC

Brian-Acosta commented on examples/Cassie/osc/osc_standing_gains.yaml:
```diff
@@ -9,33 +9,33 @@
 rows: 3
```
We don't want to merge gains to master

Brian-Acosta commented on examples/Cassie/osc/osc_walking_gains.yaml:
```diff
@@ -3,110 +3,110 @@ cols: 3
 
```
We don't want to merge these gains to master

Brian-Acosta commented on examples/Cassie/run_osc_standing_controller.cc:
```diff
@@ -278,18 +274,18 @@ int DoMain(int argc, char* argv[]) {
-  hip_yaw_right_tracking.AddJointToTrack("hip_yaw_right", "hip_yaw_rightdot");
-  osc->AddConstTrackingData(&hip_yaw_left_tracking, 0.0 * VectorXd::Ones(1));
-  osc->AddConstTrackingData(&hip_yaw_right_tracking, 0.0 * VectorXd::Ones(1));
+//  JointSpaceTrackingData hip_yaw_left_tracking(
```
We should either delete this or add a comment stating why it's commented out. (I'm in favor of just deleting it since it over-constrains the pelvis

Brian-Acosta commented on examples/Cassie/run_osc_walking_controller.cc:
```diff
@@ -12,11 +12,11 @@
 #include "multibody/kinematic/fixed_joint_evaluator.h"
 #include "multibody/kinematic/kinematic_evaluator_set.h"
 #include "multibody/multibody_utils.h"
+#include "systems/controllers/time_based_fsm.h"
```
Not sure we want to merge this either

Brian-Acosta commented on solvers/fast_osqp_solver.h:
```diff
@@ -0,0 +1,51 @@
+#pragma once
```
bit of a nitpick but it might be worth adding a comment directing people to the drake OsqpSolver class for where this came from

Brian-Acosta commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -141,6 +142,22 @@ OperationalSpaceControl::OperationalSpaceControl(
   u_min_ = u_min;
   u_max_ = u_max;
 
+  VectorXd q_min(n_v_ - 6);
```
Can we add a comment that this is setting joint limits

Brian-Acosta commented on systems/controllers/osc/operational_space_control.cc:
```diff
@@ -318,52 +347,21 @@ void OperationalSpaceControl::Build() {
-    mu_neg1 << mu_, -1;
-    mu_1 << mu_, 1;
-    one << 1;
+    MatrixXd A = MatrixXd(5, kSpaceDim);
```
Can we name this A_fcone or something and include a comment for how many faces we are using? I assume a pyramid?

Brian-Acosta reviewed (changes requested):
Looks great!  just a few files that I don't think need to be included in this merge. See other comments for small changes that would be nice to see

Brian-Acosta reviewed (approved):
LGTM - Thanks for this!
